provider/pkg/provider/previewOutputs.go

```diff
@@ -0,0 +1,218 @@
+// (i.e. they will only change if the resource is replaced). Because of this we
+// cannot blanket copy all outputs from the prior state. The ones that are not stable
+// should remain computed since they could be changing.
+//
+// Instead, we use some heuristics to determine which outputs are stable and which are not.
+// We are making the assumption that Name, Id, and Arn are stable outputs. Sometimes resources
+// will have `name`, `id`, and `arn` properties, and sometimes they will have `resourceName`, `resourceId`,
+// and `resourceArn` properties.
+func populateStableOutputs(
+	outputsFromInputs resource.PropertyMap,
+	outputsFromPriorState resource.PropertyMap,
+	readOnly []string,
+	resourceTypeName tokens.TypeName,
+) resource.PropertyMap {
+	for _, readOnlyProp := range readOnly {
+		// For now skip outputs that are part of an array property.
+		// We have no way of knowing which item in the outputs array corresponds
+		// to which item from the inputs array (since inputs could be changing).
+		if strings.Contains(readOnlyProp, "/*/") {
+			continue
+		}
+		if isStableOutput(readOnlyProp, resourceTypeName) {
+			// nested property. Sometimes the resource Arn is in a nested property
+			if strings.Contains(readOnlyProp, "/") {
+				props := strings.Split(readOnlyProp, "/")
+				current := naming.ToSdkName(props[0])
+				key := resource.PropertyKey(current)
+				if outputFromInput, ok := outputsFromInputs[key]; ok {
+					if outputFromState, ok := outputsFromPriorState[key]; ok {
+						outputsFromInputs[key] = resource.NewObjectProperty(populateStableOutputs(
+							outputFromInput.ObjectValue(),
+							outputFromState.ObjectValue(),
+							[]string{strings.Join(props[1:], "/")},
+							resourceTypeName,
+						))
+					}
+				}
+			} else {
+				key := resource.PropertyKey(naming.ToSdkName(readOnlyProp))
+				if output, ok := outputsFromPriorState[key]; ok {
+					outputsFromInputs[key] = output
+				}
+			}
+		}
+	}
+	return outputsFromInputs
+}
+
+// This calculates the outputs of a resource based on the inputs and the output
+// properties in the resource schema.
+//
+// For example, if there is a property "someProperty" that is both an input and
+// an output, the underlying type could have readonly properties meaning that part
+// of the type is an output only value. Those will not exist as input values
+// and will be marked as computed
+func previewResourceOutputs(
+	inputs resource.PropertyMap,
+	types map[string]metadata.CloudAPIType,
+	outputs map[string]schema.PropertySpec,
+	readOnly []string,
+) resource.PropertyMap {
+	result := resource.PropertyMap{}
+	for name, prop := range outputs {
+		key := resource.PropertyKey(name)
+		if inputValue, ok := inputs[key]; ok {
+			result[key] = previewOutputValue(inputValue, types, &prop.TypeSpec, filterReadOnly(readOnly, name))
+			// if the property is a readOnly property, then it's an output only value
```

Comment:
Could you explain this bit? Why are readOnly vs non-read-only values different handled differently when trying to predict outputs? One gets omitted but the other gets an Unknown value? Are those perhaps treated de-facto in the same way by the engine?

Reply:
During preview we have the inputs from the user and need to project them into outputs. We loop over all the output types and if we find an input that matches we carry over the value. If there is an output type where the value is not found in the inputs then it could be one of two cases

1. The output is a readonly type so it is not expected to be found in the inputs and is a computed value.
2. The output is not a readonly type which means that it is an input, but the user has just not specified it.

For these cases I assumed that we should just leave it as omitted, but should we just mark these as computed too? I suppose some of them could be fields that have default values which means that the value would be computed. The CCAPI schema isn't verbose enough to tell us whether an input is computed or not so should we just assume it is?

Reply:
Ah that's a helpful. I think your code does the right thing. I suspect omitting something is treated similar to having it as Null, but sending an Unknown is different from Null as it signals something might change during update. 

1. The output is a readonly type so it is not expected to be found in the inputs and is a computed value.
This case should be unknown. If we leave it undefined (null), we're being over-confident in that the provider will not specify it.

2. The output is not a readonly type which means that it is an input, but the user has just not specified it.
This case should be left undefined. We are being confident here that the provider will not modify this.

Reply:
We do nothing with default values. If there is an input/output property call `foo` and the user does not populate it then the input will be empty. The output though is populated as a result of reading from CCAPI so if AWS returns a default value the output will have that default value.

Reply:
I opted to leave it as is. If we mark every output property as computed it could make the preview very verbose since properties that might not be populated will show as computed.